runtime/src/current_sandbox.c

```diff
@@ -88,20 +103,43 @@ current_sandbox_start(void)
 	current_sandbox_disable_preemption(sandbox);
 	sandbox->completion_timestamp = __getcycles();
 
-	/* Retrieve the result, construct the HTTP response, and send to client */
-	if (sandbox_send_response(sandbox) < 0) {
-		error_message = "Unable to build and send client response\n";
-		goto err;
-	};
 
-	http_total_increment_2xx();
+	if (next_module != NULL) {
+		/* Generate a new request, copy the current sandbox's output to the next request's buffer, and put it to the global queue */
+		ssize_t output_length = sandbox->request_response_data_length - sandbox->request_length;
+		char * pre_func_output = (char *) malloc(output_length);
+		memcpy(pre_func_output, sandbox->request_response_data + sandbox->request_length, output_length);
+	 	uint64_t enqueue_timestamp = __getcycles();	
+		struct sandbox_request *sandbox_request =
+                                  sandbox_request_allocate(next_module, false, sandbox->request_length, 
+							   next_module->name, sandbox->client_socket_descriptor,
+                                                           (const struct sockaddr *)&sandbox->client_address,
+                                                           sandbox->request_arrival_timestamp, enqueue_timestamp, 
+							   true, pre_func_output, output_length);
+		/* TODO: all sandboxs in the chain share the same request id, but sandbox_request_allocate() will busy-wait to generate an unique
+		   id, should we optimize it here?*/
+		sandbox_request->id = sandbox->id;  
+		/* Add to the Global Sandbox Request Scheduler */
+		global_request_scheduler_add(sandbox_request);
+		if (sandbox->request_from_outside) {
+			sandbox_remove_from_epoll(sandbox);
```

Comment:
I think that I'm missing something there.

As we never returned from `accept` to trigger this sandbox, we didn't create a new `fd` for it. So that was never added to the `epoll` set. Is this just removing the initial `fd` used to access the sandbox from the network from the epoll set? If so, then we should *not* be doing it for each activation of the sandbox (this is not just a performance problem, you could be removing `fd`s that belong to other sandboxes). So if this is removing the `fd` for the overall module, then I'd expect it would be done 1. at bootup time when we're parsing the modules, and 2. should also include closing the fd. Thus, it would be done only once.

I get the feeling here that I'm missing something. I think this is an important discussion. @bushidocodes @emil916

Reply:
Check return value? I'm guessing this is trying to remove `0` from the epoll set, which should error out?

Reply:
> I think that I'm missing something there.
> 
> As we never returned from `accept` to trigger this sandbox, we didn't create a new `fd` for it. So that was never added to the `epoll` set. Is this just removing the initial `fd` used to access the sandbox from the network from the epoll set? If so, then we should _not_ be doing it for each activation of the sandbox (this is not just a performance problem, you could be removing `fd`s that belong to other sandboxes). So if this is removing the `fd` for the overall module, then I'd expect it would be done 1. at bootup time when we're parsing the modules, and 2. should also include closing the fd. Thus, it would be done only once.
> 
> I get the feeling here that I'm missing something. I think this is an important discussion. @bushidocodes @emil916

You are right, the return of accept won't trigger the sandbox. The fd here is the return value of accept(), it is the client fd. When a sandbox is running, it will call "sandbox_open_http", within this function, it will add this client fd to the epoll to monitor its IN/OUT/ERR event because sledge will do read/write from/to this fd and it can be blocked. If it is the first module (sandbox->request_from_outside is true if it is the first module), sledge should remove the client fd from epoll but not close if before the sandbox exits, otherwise, the sandbox is freed and epoll still return the client fd, but epoll_event.data.ptr is invalid because the sandbox is freed(epoll_event.data.ptr = (void *)sandbox), so we must remove it here when the sandbox finished.

Reply:
> Check return value? I'm guessing this is trying to remove `0` from the epoll set, which should error out?

Yes, should check the error. It's inside of `the sandbox_remove_from_epoll `
```
static inline void
sandbox_remove_from_epoll(struct sandbox *sandbox)
{
        assert(sandbox != NULL);

        int rc = epoll_ctl(worker_thread_epoll_file_descriptor, EPOLL_CTL_DEL, sandbox->client_socket_descriptor, NULL);
        if (unlikely(rc < 0)) panic_err();

}
```

Reply:
Thanks for the input! I was confused, and thought the epoll delete was done during IPC, but you're right: it is done at initialization. I want to confirm that these goals are maintained:

- no sockets are created for modules that are not outward facing.
- Only outward-facing modules have sockets, and `accept` returns a *single* client's fd.
- the socket used to communicate with the client is both used to read the input, and to send the output *after* the last sandbox in the chain executes.

I'm hesitant to believe that these goals are met as it seems that the fds are made, added, then removed.

Reply:
I should also note: if the fd is created for modules in a chain, currently, that's likely OK at the current point in time. Once you add the json support for chains to form them along-side other modules, then it will be a good time to fix which modules have external sockets and which dont.

Reply:
Modules only create and epoll monitor the socket fds for the server-side(listen the socket fd to wait for the requests coming in), this is done when a module was added. The listening thread will do `accept()` and return a client-side fd, this client fd is monitored by epoll, read/write data within a sandbox.